### ISSUE TITLE
[Feature Request]: Extend Code Injection Capabilities for PsychoPy Builder

### ISSUE THREAD

**Issue by wieluk** [psychopy/psychopy#7005]
### Description

PsychoPy allows plugins to inject code during experiment generation by writing from the Builder into the .py file. It also supports plugins with hooks that can insert code into specific functions like writeStartCode and writeExperimentEndCode(for Routines). However, these hooks are limited and do not allow injecting code into other functions such as saveData or endExperiment. This request proposes extending the hook system to enable injection into these functions, providing more flexibility for developers.

### Purpose

Extending the hook system will allow developers to customize behavior during runtime, such as copying generated data to specific directories or adding post-experiment cleanup tasks, directly from their plugins.

### Is your feature request related to a problem?

While PsychoPy’s plugin hooks are helpful, their scope is limited. Developers cannot influence functions like data saving or experiment ending directly through the current hook system.

### Describe the solution you would like

Add hooks in `BaseStandaloneRoutine` for saveData and endExperiment functions, allowing plugins to inject code into these stages of the experiment. This would expand the flexibility of PsychoPy plugins.

**Comment by peircej** [comment]
This sounds appropriate to me. Is there anything I'm missing @TEParsons @mdcutone ?

**Comment by TEParsons** [comment]
Yep this seems doable! I think we just need for loops at the end of `SettingsComponent.writeEndCode` and `SettingsComponent.writeSaveDataCode` similar to what we have in `SettingsComponent.writeDevicesCode`
